primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -0,0 +1,731 @@
+	}
+
+	/// Get a mutable reference for a value.
+	///
+	/// Can be rolled back or committed when called inside a transaction.
+	#[must_use = "A change was registered, so this value MUST be modified."]
+	pub fn modify(
+		&mut self,
+		key: StorageKey,
+		init: impl Fn() -> StorageValue,
+		at_extrinsic: Option<u32>
+	) -> &mut Option<StorageValue> {
+		let overlayed = self.changes.entry(key.clone()).or_insert_with(Default::default);
+		let first_write_in_tx = insert_dirty(&mut self.dirty_keys, key);
+		let clone_into_new_tx = if let Some(tx) = overlayed.transactions.last() {
+			if first_write_in_tx {
+				Some(tx.value.clone())
+			} else {
+				None
+			}
+		} else {
+			Some(Some(init()))
+		};
+
+		if let Some(cloned) = clone_into_new_tx {
+			overlayed.set(cloned, first_write_in_tx, at_extrinsic);
+		}
+		overlayed.value_mut()
+	}
+
+	/// Set all values to deleted which are matched by the predicate.
+	///
+	/// Can be rolled back or committed when called inside a transaction.
+	pub fn clear_where(
+		&mut self,
+		predicate: impl Fn(&[u8], &OverlayedValue) -> bool,
+		at_extrinsic: Option<u32>
+	) {
+		for (key, val) in self.changes.iter_mut().filter(|(k, v)| predicate(k, v)) {
+			val.set(None, insert_dirty(&mut self.dirty_keys, key.to_owned()), at_extrinsic);
+		}
+	}
+
+	/// Get a list of all changes as seen by current transaction.
+	pub fn changes(&self) -> impl Iterator<Item=(&StorageKey, &OverlayedValue)> {
+		self.changes.iter()
+	}
+
+	/// Get the change that is next to the supplied key.
+	pub fn next_change(&self, key: &[u8]) -> Option<(&[u8], &OverlayedValue)> {
+		use std::ops::Bound;
+		let range = (Bound::Excluded(key), Bound::Unbounded);
+		self.changes.range::<[u8], _>(range).next().map(|(k, v)| (&k[..], v))
+	}
+
+	/// Consume this changeset and return all committed changes.
+	///
+	/// Panics:
+	/// Panics if there are open transactions: `transaction_depth() > 0`
+	pub fn drain_commited(self) -> impl Iterator<Item=(StorageKey, Option<StorageValue>)> {
+		assert!(self.transaction_depth() == 0, "Drain is not allowed with open transactions.");
+		self.changes.into_iter().map(|(k, mut v)| (k, v.pop_transaction().value))
+	}
+
+	/// Returns the current nesting depth of the transaction stack.
+	///
+	/// A value of zero means that no transaction is open and changes are committed on write.
+	pub fn transaction_depth(&self) -> usize {
+		self.dirty_keys.len()
+	}
+
+	/// Returns how many of the open transactions are opened by the client.
+	///
+	/// These are alwyays the first transaction to open and the last to close.
+	pub fn num_client_transactions(&self) -> usize {
+		self.num_client_transactions
+	}
+
+	/// Returns whether we are executing in client or runtime mode.
+	pub fn execution_mode(&self) -> ExecutionMode {
+		self.execution_mode
+	}
+
+	/// Call this before entering runtime.
+	pub fn enter_runtime(&mut self) {
+		if let ExecutionMode::Runtime = self.execution_mode {
+			return;
+		}
```

Comment:
TBH, I am not sure if ignoring the second `enter_runtime` call is a good idea - we expect only a single `enter_runtime`, right? OTOH, I don't see a way to mess up anything with that rn. If we keep this as is, let's leave a word about this behavior in the docs?

Reply:
A second `enter_runtime` after already starting some transactions inside the runtime would be bad as then we would be writing something wrong to `num_client_transactions`. Therefore I prevent that.

Of course we can write to the docs that a second `enter` or a second `exit` for that matter is is a no-op. On `exit` that happens naturally.